src/platform/nodejs/persist/fileSystemPersist.ts

```diff
@@ -0,0 +1,128 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import { Logger } from '@opentelemetry/api';
+import { NoopLogger } from '@opentelemetry/core';
+import { PersistentStorage } from '../../../types';
+import { ExporterConfig, DEFAULT_EXPORTER_CONFIG } from '../../../config';
+import { confirmDirExists, getShallowDirectorySize } from './fileSystemHelpers';
+
+export class FileSystemPersist implements PersistentStorage {
+  static TEMPDIR_PREFIX = 'ot-azure-exporter-';
+
+  maxBytesOnDisk: number = 50_000_000; // ~50MB
+
+  private readonly _options: ExporterConfig;
+
+  private readonly _logger: Logger;
+
+  constructor(options: Partial<ExporterConfig> = DEFAULT_EXPORTER_CONFIG) {
+    this._options = { ...DEFAULT_EXPORTER_CONFIG, ...options };
+    this._logger = options.logger || new NoopLogger();
+    if (!this._options.instrumentationKey) {
+      this._logger.error(
+        `No instrumentation key was provided to FileSystemPersister. Files may not be properly persisted`,
+      );
+    }
+  }
+
+  push(value: unknown, cb: (err: Error | null, result?: boolean | undefined) => void): void {
+    this._logger.info('Pushing value to persistent storage', value as string);
+    this._storeToDisk(JSON.stringify(value), cb);
+  }
+
+  shift(cb: (err: Error | null, value?: unknown) => void): void {
+    this._logger.info('Returning first member of filesystem');
+    this._getFirstFileOnDisk((error, buffer) => {
+      if (error) {
+        cb(error);
+      } else if (buffer) {
+        cb(null, JSON.parse(buffer.toString('utf8')));
+      }
+    });
+  }
+
+  /**
+   * Check for temp telemetry files
+   * reads the first file if exist, deletes it and tries to send its load
+   */
+  private _getFirstFileOnDisk(callback: (error: Error | null, value?: Buffer) => void): void {
+    const tempDir = path.join(
+      os.tmpdir(),
+      FileSystemPersist.TEMPDIR_PREFIX + this._options.instrumentationKey,
+    );
+
+    fs.stat(tempDir, (statErr: Error | null, stats: fs.Stats) => {
+      if (stats.isDirectory()) {
+        fs.readdir(tempDir, (error, origFiles) => {
+          if (!error) {
+            const files = origFiles.filter((f) => path.basename(f).includes('.ai.json'));
+            if (files.length > 0) {
+              const firstFile = files[0];
+              const filePath = path.join(tempDir, firstFile);
+              fs.readFile(filePath, (readFileErr, payload) => {
+                if (!readFileErr) {
+                  // delete the file first to prevent double sending
+                  fs.unlink(filePath, (unlinkError) => {
```

Comment:
When would double sending occur (if we didn't delete this file)?

Reply:
It is mainly a race condition when we try to send a file and then the server shuts down while waiting for a response. We either assume it was sent successfully and delete it or double send it.  Previous node.js SDK behavior is to delete first, so I'm just keeping that here. How does python handle this scenario @lzchen?

Reply:
Why do you need to assume it's sent successfully or not? If there is no response sent back from the server, it is treated as another retryable case: the file is left in storage to be retried again in the future. Upon successful response from server, then the file is deleted.

Reply:
envelope is sent --> client shuts down --> response is never handled

Reply:
Oh currently Python does not handle this race condition. there is no way to tell whether the server accepted the data or not at that very moment.